tests/cpp/test_multidevice_tutorial.cpp

```diff
@@ -0,0 +1,367 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <multidevice/communicator.h>
+#include <ops/all_ops.h>
+#include <tests/cpp/multidevice.h>
+
+namespace nvfuser {
+
+class MultiDeviceTutorial : public MultiDeviceTest {
+ protected:
+  static void SetUpTestSuite() {
+    verbose_ = getNvFuserEnv("TUTORIAL_VERBOSE");
+  }
+
+ protected:
+  static bool verbose_;
+};
+
+bool MultiDeviceTutorial::verbose_ = false;
+
+// The first object we need to introduce is the class `Communicator` which is
+// a convenience that is nvFuser's interface to runtime distributed setting
+// and backend. Communicator's setup is expansive (because all IP addresses
+// need to be exchanged to all ranks) and occupies a port. Therefore, we
+// instantiate one Communicator globally and use it accross all tests. There
+// is no benefit in having multiple communicator, and this class is thought to
+// have a singleton instance. This communicator can be accessed through
+// MultiDeviceTest::communicator_
+TEST_F(MultiDeviceTutorial, CommunicatorAndC10d) {
+  Communicator* communicator = communicator_;
+
+  // To check if distributed setting is available:
+  if (communicator->is_available() == false) {
+    GTEST_SKIP() << "distributed setting not available";
+  }
+  // During instantiation, the Communicator will parse information from the
+  // environment variable. Setting those env vars properly is what is requested
+  // from the user for multidevice to be available in nvFuser. Those environment
+  // variables are set automatically in the case of a single-node system where
+  // "mpirun" (from open MPI) is used to launch the application. The actual list
+  // of required environment variables are:
+  //  - Set automatically by mpirun (so, no need to manually set them if using
+  //  mpirun):
+  //      - OMPI_COMM_WORLD_RANK
```

Comment:
Can we use other MPI implementations that may not define `OMPI_` environment variables?

Reply:
We actually also check the same variable without the `OMPI_COMM_` prefix, check https://github.com/NVIDIA/Fuser/blob/6866230b4c4417ffefed5de8338badc4e966766b/csrc/multidevice/communicator.cpp#L59
This is supposed to be how torchrun sets env vars, so we theoretically also support torch run, and I can't forsee any issue using this alternative ; but since we don't test torchrun, it is safer to always use mpirun.

Anyway I though it was not worth mentionning this alternative in the tutorial, but lmk if you think otherwise

Reply:
I was thinking of other MPI implementations such as MPICH since (annoyingly and IIRC) these MPI implementation failed to agree on environment names and thus use slightly different ones.

It's ok to just support OpenMPI but it should be mentioned somewhere.

Reply:
The exact requirement is to have those envs set properly. One sufficient (but not necessary) condition is to use ompi -- if using something else, one needs to make sure those env vars are still setup properly.

I think it is clearly conveyed in the present comment but lmk your thoughts